### ISSUE TITLE
Floating point values not being written correctly

### ISSUE THREAD

**Issue by jazzay** [wapc/as-msgpack#4]
The library currently contains this implementation:
```
setFloat32(value: f32): void {
    if (this.byteOffset + 4 > this.byteLength)
      throw new RangeError(E_INDEXOUTOFRANGE);
    store<u32>(this.dataStart + this.byteOffset, <u32>reinterpret<u32>(value));
    this.byteOffset += 4;
  }
```
Why the reinterpret?
When I fix that locally and then try to read from rust I get incorrect f32 values coming across. :(

**Comment by pkedy** [comment]
Hi @jazzay! Can you elaborate on "when I fix that locally"?  Does that mean you removed it?  Is there an issue with `reinterpret`?  The intent is to "reinterpret" f32 as u32 for serialization purposes. The AS doc site says this about `reinterpret`.

From https://www.assemblyscript.org/environment.html#low-level-webassembly-operations:
> Reinterprets the bits of the specified value as type T. Valid reinterpretations are u32/i32 to/from f32 and u64/i64 to/from f64.

I pushed some tests (work in progress) [here](https://github.com/wapc/language-tests). For each language, including AS, the [F32 and F64 tests](https://github.com/wapc/language-tests/blob/master/pkg/module/module_test.go#L56-L57) are currently passing. Do you see something in there that is not correct?

Also, thanks for submitting this issue. Getting serialization perfect is a huge concern! The language-tests repo is a step toward making sure all of our code generation and MsgPack tooling works as expected across all languages. Next step is to add fuzzing. Let me know if you have any ideas for making it better. Thanks!

**Comment by jazzay** [comment]
No problem! I changed it to this:
```
  setFloat32(value: f32): void {
    if (this.byteOffset + 4 > this.byteLength)
      throw new RangeError(E_INDEXOUTOFRANGE);
	store<f32>(this.dataStart + this.byteOffset, value);   // <---- changed
    this.byteOffset += 4;
  }
```
Using the native AS store operator. Not sure why it is coming back invalid on the rust host side.
Will try to dig into that and understand why...

**Comment by MaxGraey** [comment]
I guess it's because with `store<f32>` input float values for `-0.0` normalized to `0.0`. Same for non-canonical NaNs while `store<u32>(..., reinterpret<u32>(value))` store float number bit to bit. However I'm not sure about zeros but NaNs definitely will be canonicalized during store operation.

**Comment by jazzay** [comment]
I guess maybe it's an issue on the rust side then with decoding those floats?
Do you plan to implement your test host/harness in rust as well?

**Comment by pkedy** [comment]
@jazzay I think we'll implement it in Rust eventually just to make sure the `wapc-rust` host is covered. What's funny to me is that the Rust guest is echoing back the float values which means Rust (at least in the Wasm target) is decoding it properly. Any ideas @autodidaddict ?

**Comment by jazzay** [comment]
Possibly it could be a byte ordering issue? After digging into msg pack for rust (rmp) I see that it needs to write in Big endian, while Wasm memory itself is little. I do not see any byte order swapping in the as-msgpack code base. Echoing it back possibly hides this issue?

**Comment by pkedy** [comment]
@jazzay Good point and very likely. I'll look into that and report back.

**Comment by pkedy** [comment]
@jazzay `1234` from the test becomes `5.932811833064553e-39` in AS so I think you are correct!

**Comment by pkedy** [comment]
@jazzay please give this a try.  @MaxGraey please let me know if there is a cleaner way.

```ts
  getFloat32(): f32 {
    if (this.byteOffset + 4 > this.byteLength)
      throw new RangeError(E_INDEXOUTOFRANGE);
    const result = reinterpret<f32>(bswap(load<u32>(this.dataStart + this.byteOffset)));
    this.byteOffset += 4;
    return result;
  }

  getFloat64(): f64 {
    if (this.byteOffset + 8 > this.byteLength)
      throw new RangeError(E_INDEXOUTOFRANGE);
    const result = reinterpret<f64>(bswap(load<u64>(this.dataStart + this.byteOffset)));
    this.byteOffset += 8;
    return result;
  }
```

```ts
  setFloat32(value: f32): void {
    if (this.byteOffset + 4 > this.byteLength)
      throw new RangeError(E_INDEXOUTOFRANGE);
    store<u32>(this.dataStart + this.byteOffset, bswap(reinterpret<u32>(value)));
    this.byteOffset += 4;
  }

  setFloat64(value: f64): void {
    if (this.byteOffset + 8 > this.byteLength)
      throw new RangeError(E_INDEXOUTOFRANGE);
    store<u64>(this.dataStart + this.byteOffset, bswap(reinterpret<u64>(value)));
    this.byteOffset += 8;
  }
```

**Comment by MaxGraey** [comment]
Yes, looks good. Just realized MessagePack is big endian by design due to most network protocols also have BE order.

**Comment by jazzay** [comment]
@pkedy I tested out your submitted fix, and it solves my issue.
Thanks for turning that around so quickly!

On a side note, are you considering alternate encoding protocols (little endian) that avoid that overhead?
I took a look at flat buffers for example, but the typescript implementation does not work in AS currently.

**Comment by pkedy** [comment]
@jazzay No problem.  I'm going to add decoding tests to cover these cases.  After that I'll be creating a release of this library and also pushing to npm.  Just hang on using the master branch in the meantime.

Yes, we will be looking into alternate serialization formats.  We are aware of Protobuf use cases so we'll probably go there next.  That said, we probably only want to support a few formats as each one adds effort to maintain. We choose MsgPack as the primary format because we found it to be faster (than Protobuf) and has the added benefit in that it can be parsed without a schema which is useful for several use cases. We want to optimize for speed over compression because in _most scenarios_ the data does not go over the wire.  I looked at FlatBuffers a while back but I'd have to refresh myself on how it works at a low level.
